File: langs/StdLibs/one.py
import re

class Regex:
    @staticmethod
    def matchFromIndex(pattern, input, offset):
        patternObj = re.compile(pattern)
        match = patternObj.match(input, offset)
        if not match:
            return None
        return list(match.group(0)) + list(match.groups())

class Reflect:
    classes = {}

    @staticmethod
    def get_class(obj):
        return Reflect.classes.get(Reflect.name_key(obj.__class__.__name__))

    @staticmethod
    def get_class_by_name(name):
        return Reflect.classes.get(Reflect.name_key(name))

    @staticmethod
    def setup_class(cls):
        Reflect.classes[Reflect.name_key(cls.name)] = cls

    @staticmethod
    def name_key(name):
        return name.lower().replace("_", "")

class Class:
    def __init__(self, typeObj, fields, methods):
        self.typeObj = typeObj
        self.name = typeObj.__name__
        
        self.fields = {}
        for field in fields:
            field.cls = self
            self.fields[Reflect.name_key(field.name)] = field
        
        self.methods = {}
        for method in methods:
            method.cls = self
            self.methods[Reflect.name_key(method.name)] = method
    
    def get_field(self, name):
        return self.fields.get(Reflect.name_key(name))
        
    def get_method(self, name):
        return self.methods.get(Reflect.name_key(name))

    def get_fields(self):
        return self.fields.values()

    def get_methods(self):
        return self.methods.values()

class Field:
    def __init__(self, name, is_static):
        self.cls = None
        self.name = name
        self.is_static = is_static
    
    def get_value(self, obj):
        realObj = self.cls.typeObj if self.is_static else obj
        return getattr(realObj, self.name)
    
    def set_value(self, obj, value):
        realObj = self.cls.typeObj if self.is_static else obj
        setattr(realObj, self.name, value)

class Method:
    def __init__(self, name, is_static, args):
        self.cls = None
        self.name = name
        self.is_static = is_static
        self.args = args
    
    def call(self, obj, args):
        realObj = self.cls.typeObj if self.is_static else obj
        method = getattr(realObj, self.name)
        return method(*args)

class MethodArgument:
    def __init__(self, name, type):
        self.name = name
        self.type = type

File: langs/StdLibs/one.js
class Regex {
    static matchFromIndex(pattern, input, offset) {
        const regex = new RegExp(pattern, "gy");
        regex.lastIndex = offset;
        const matches = regex.exec(input);
        return matches === null ? null : Array.from(matches);
    }
}

class Reflect {
    static getClass(obj) { return this.classes[obj.constructor.name.toLowerCase()]; }
    static getClassByName(name) { return this.classes[name.toLowerCase()]; }

    static setupClass(cls) { this.classes[cls.name.toLowerCase()] = cls; }
}

Reflect.classes = {};

class Class {
    constructor(typeObj, fields, methods) {
        this.typeObj = typeObj;
        this.name = typeObj.name;
        this.fields = {};
        this.methods = {};

        for (const field of fields) {
            field.cls = this;
            this.fields[field.name.toLowerCase()] = field;
        }

        for (const method of methods) {
            method.cls = this;
            this.methods[method.name.toLowerCase()] = method;
        }
    }

    getField(name) { return this.fields[name.toLowerCase()]; }
    getMethod(name) { return this.methods[name.toLowerCase()]; }

    getFields() { return Object.values(this.fields); }
    getMethods() { return Object.values(this.methods); }
}

class Field {
    constructor(name, isStatic, type) {
        this.name = name;
        this.isStatic = isStatic;
        this.type = type;
    }

    getValue(obj) {
        const realObj = this.isStatic ? this.cls.typeObj : obj;
        return realObj[this.name];
    }

    setValue(obj, value){
        const realObj = this.isStatic ? this.cls.typeObj : obj;
        realObj[this.name] = value;
    }
}

class Method {
    constructor(name, isStatic, returnType, args) {
        this.name = name;
        this.isStatic = isStatic;
        this.args = args;
        this.returnType = returnType;
    }

    call(obj, args) {
        if (args.length !== this.args.length)
            throw new Error(`Expected ${this.args.length} arguments, but got ${args.length} in ${this.cls.name}::${this.name} call!`);

        const realObj = this.isStatic ? this.cls.typeObj : obj;
        const method = realObj[this.name]
        return method.apply(obj, args);
    }
}

class MethodArgument {
    constructor(name, type) {
        this.name = name;
        this.type = type;
    }
}

module.exports = { Regex, Reflect, Class, Field, Method, MethodArgument };